hadoop-ozone/client/src/test/java/org/apache/hadoop/ozone/client/TestBlockOutputStreamIncrementalPutBlock.java

```diff
@@ -0,0 +1,156 @@
+  private OzoneBucket bucket;
+  private final ConfigurationSource config = new InMemoryConfiguration();
+
+  public static Iterable<Boolean> parameters() {
+    return Arrays.asList(true, false);
+  }
+
+  public void init(boolean incrementalChunkList) throws IOException {
+    OzoneClientConfig clientConfig = config.getObject(OzoneClientConfig.class);
+
+    clientConfig.setIncrementalChunkList(incrementalChunkList);
+    clientConfig.setChecksumType(ContainerProtos.ChecksumType.CRC32C);
+
+    ((InMemoryConfiguration)config).setFromObject(clientConfig);
+
+    RpcClient rpcClient = new RpcClient(config, null) {
+
+      @Override
+      protected OmTransport createOmTransport(
+          String omServiceId)
+          throws IOException {
+        return new MockOmTransport();
+      }
+
+      @NotNull
+      @Override
+      protected XceiverClientFactory createXceiverClientFactory(
+          ServiceInfoEx serviceInfo) throws IOException {
+        return new MockXceiverClientFactory();
+      }
+    };
+
+    client = new OzoneClient(config, rpcClient);
+    ObjectStore store = client.getObjectStore();
+
+    store.createVolume(volumeName);
+    OzoneVolume volume = store.getVolume(volumeName);
+    volume.createBucket(bucketName);
+    bucket = volume.getBucket(bucketName);
+  }
+
+  @AfterEach
+  public void close() throws IOException {
+    client.close();
+  }
+
+  @ParameterizedTest
+  @MethodSource("parameters")
+  public void writeSmallKey(boolean incrementalChunkList)
+      throws IOException {
+    init(incrementalChunkList);
+
+    int size = 1024;
+    String s = RandomStringUtils.randomAlphabetic(1024);
+    ByteBuffer byteBuffer = ByteBuffer.wrap(s.getBytes(StandardCharsets.UTF_8));
+
+    try (OzoneOutputStream out = bucket.createKey(keyName, size,
+        ReplicationConfig.getDefault(config), new HashMap<>())) {
+      for (int i = 0; i < 4097; i++) {
+        out.write(byteBuffer);
+        out.hsync();
+      }
+    }
+
+    try (OzoneInputStream is = bucket.readKey(keyName)) {
+      ByteBuffer readBuffer = ByteBuffer.allocate(size);
+      for (int i = 0; i < 4097; i++) {
+        is.read(readBuffer);
+        assertArrayEquals(readBuffer.array(), byteBuffer.array());
+      }
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("parameters")
+  public void writeLargeKey(boolean incrementalChunkList)
+      throws IOException {
+    init(incrementalChunkList);
+
+    int size = 1024 * 1024 + 1;
+    ByteBuffer byteBuffer = ByteBuffer.allocate(size);
+
+    try (OzoneOutputStream out = bucket.createKey(keyName, size,
+        ReplicationConfig.getDefault(config), new HashMap<>())) {
+      for (int i = 0; i < 4; i++) {
+        out.write(byteBuffer);
+        out.hsync();
+      }
+    }
+
+    try (OzoneInputStream is = bucket.readKey(keyName)) {
+      ByteBuffer readBuffer = ByteBuffer.allocate(size);
+      for (int i = 0; i < 4; i++) {
+        is.read(readBuffer);
```

Comment:
I added one check here, it failed

`
   int readLen = is.read(readBuffer);
   assertEquals(bufferSize, readLen);
`